ssi-sd-jwt/src/encode.rs

```diff
@@ -0,0 +1,218 @@
+use jose_b64::base64ct::{Base64UrlUnpadded, Encoding};
+use rand::{CryptoRng, Rng};
+use serde::Serialize;
+use ssi_jwk::{Algorithm, JWK};
+
+use crate::*;
+
+fn encode_disclosure_with_salt<ClaimValue: Serialize>(
+    salt: &str,
+    claim_name: Option<&str>,
+    claim_value: &ClaimValue,
+) -> Result<String, serde_json::Error> {
+    let disclosure = match claim_name {
+        Some(claim_name) => serde_json::json!([salt, claim_name, claim_value]),
+        None => serde_json::json!([salt, claim_value]),
+    };
+
+    let json_bytes = jose_b64::serde::Json::<serde_json::Value>::new(disclosure)?;
+
+    Ok(Base64UrlUnpadded::encode_string(json_bytes.as_ref()))
+}
+
+pub fn encode_disclosure_with_rng<ClaimValue: Serialize, Rand: Rng + CryptoRng>(
+    rng: &mut Rand,
+    claim_name: Option<&str>,
+    claim_value: &ClaimValue,
+) -> Result<String, serde_json::Error> {
+    // TODO: link to rfc wrt suggested bit size of salt
+    const DEFAULT_SALT_SIZE: usize = 128 / 8;
+
+    let mut salt_bytes = [0u8; DEFAULT_SALT_SIZE];
+
+    rng.fill_bytes(&mut salt_bytes);
+
+    let salt = Base64UrlUnpadded::encode_string(&salt_bytes);
+
+    encode_disclosure_with_salt(&salt, claim_name, claim_value)
+}
+
+fn encode_disclosure<ClaimValue: Serialize>(
+    claim_name: Option<&str>,
+    claim_value: &ClaimValue,
+) -> Result<String, serde_json::Error> {
+    let mut rng = rand::rngs::OsRng {};
+    encode_disclosure_with_rng(&mut rng, claim_name, claim_value)
+}
+
+pub fn encode_property_disclosure<ClaimValue: Serialize>(
+    sd_alg: SdAlg,
+    claim_name: &str,
+    claim_value: &ClaimValue,
+) -> Result<Disclosure, serde_json::Error> {
+    let encoded = encode_disclosure(Some(claim_name), claim_value)?;
+    let hash = hash_encoded_disclosure(sd_alg, &encoded);
+
+    Ok(Disclosure { encoded, hash })
+}
+
+pub fn encode_array_disclosure<ClaimValue: Serialize>(
+    sd_alg: SdAlg,
+    claim_value: &ClaimValue,
+) -> Result<Disclosure, serde_json::Error> {
+    let encoded = encode_disclosure(None, claim_value)?;
+    let hash = hash_encoded_disclosure(sd_alg, &encoded);
+
+    Ok(Disclosure { encoded, hash })
+}
+
+pub fn encode_sign<Claims: Serialize>(
```

Comment:
The main issue is that the return value array of encoded disclosures were very difficult to match up to the various disclosures, since very complex sets of disclosures are possible including nesting.  True, the object is not used much from the encode side except for the base jwt (so common claims like iss, exp, etc. that will always be disclosed), but in exchange you do get to keep the same claim struct on the decode and encode side.